packages/gitbook/src/app/middleware/(site)/(core)/sitemap.xml/route.ts

```diff
@@ -68,30 +56,66 @@ export async function GET(req: NextRequest) {
     });
 }
 
-type FlatPageEntry = { page: RevisionPageDocument; depth: number };
-
-function flattenPages(
-    rootPags: RevisionPage[],
-    filter: (page: RevisionPageDocument | RevisionPageGroup) => boolean,
-): FlatPageEntry[] {
-    const flattenPage = (
-        page: RevisionPageDocument | RevisionPageGroup,
-        depth: number,
-    ): FlatPageEntry[] => {
-        const allowed = filter(page);
-        if (!allowed) {
-            return [];
+/**
+ * Check if the pointer is the root one.
+ * Meaning we are on the default section / space.
+ */
+function checkIsRootPointer(pointer: SiteContentPointer, siteStructure: SiteStructure): boolean {
+    switch (siteStructure.type) {
+        case 'sections': {
+            return siteStructure.structure.some(
+                (structure) =>
+                    structure.default &&
+                    structure.id === pointer.siteSectionId &&
+                    structure.siteSpaces.some(
+                        (siteSpace) => siteSpace.default && siteSpace.id === pointer.siteSpaceId,
+                    ),
+            );
+        }
+        case 'siteSpaces': {
+            return siteStructure.structure.some(
+                (siteSpace) => siteSpace.default && siteSpace.id === pointer.siteSpaceId,
+            );
         }
+    }
+}
 
-        return [
-            ...(page.type === 'document' ? [{ page, depth }] : []),
-            ...page.pages.flatMap((child) =>
-                child.type === 'document' ? flattenPage(child, depth + 1) : [],
-            ),
-        ];
-    };
+/**
+ * Get Sitemap URLs from site structure.
+ */
+async function getUrlsFromSiteStructure(siteStructure: SiteStructure): Promise<string[]> {
+    switch (siteStructure.type) {
+        case 'sections':
+            return getUrlsFromSiteSections(siteStructure.structure);
+        case 'siteSpaces':
+            return getUrlsFromSiteSpaces(siteStructure.structure);
+        default:
+            assertNever(siteStructure);
+    }
+}
+
+/**
+ * Get Sitemap URLs from site sections.
+ */
+async function getUrlsFromSiteSections(siteSections: SiteSection[]): Promise<string[]> {
+    const urls = await Promise.all(
+        siteSections.map(async (siteSection) => getUrlsFromSiteSpaces(siteSection.siteSpaces), []),
+    );
+    return urls.flat();
+}
 
-    return rootPags.flatMap((page) =>
-        page.type === 'group' || page.type === 'document' ? flattenPage(page, 0) : [],
+/**
+ * Get Sitemap URLs from site spaces.
+ */
+async function getUrlsFromSiteSpaces(siteSpaces: SiteSpace[]): Promise<string[]> {
+    const urls = await Promise.all(
+        siteSpaces.map(async (siteSpace) => {
+            if (!siteSpace.urls.published) {
+                return null;
+            }
+            const url = new URL(siteSpace.urls.published);
+            return getAbsoluteHref(joinPath(url.pathname, 'sitemap-pages.xml'), true);
```

Comment:
I'm not sure this logic will work in all cases:

- `getAbsoluteHref` will append based on the current request URL
- if a site is served under a slug in a org domain (ex: orgdomain.com/siteslug`)
- a part of the URL will be lost
```suggestion
            const url = new URL(siteSpace.urls.published);
            url.pathname = joinPath(url.pathname, 'sitemap-pages.xml')
            return url.toString();
```